istioctl/cmd/tag.go

```diff
@@ -0,0 +1,400 @@
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/spf13/cobra"
+	admit_v1 "k8s.io/api/admissionregistration/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"istio.io/api/label"
+)
+
+const (
+	// TODO(monkeyanator) move into istio/api
+	istioTagLabel             = "istio.io/tag"
+	istioInjectionWebhookName = "sidecar-injector.istio.io"
+	revisionTagNamePrefix     = "istio-revision-tag"
+)
+
+var (
+	// revision to point tag webhook at
+	revision         = ""
+	overwrite        = false
+	skipConfirmation = false
+)
+
+func tagCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "tag",
+		Short: "Command group used to interact with revision tags",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cmd.HelpFunc()(cmd, args)
+			if len(args) != 0 {
+				return fmt.Errorf("unknown subcommand %q", args[0])
+			}
+
+			return nil
+		},
+	}
+
+	cmd.AddCommand(tagSetCommand())
+	cmd.AddCommand(tagListCommand())
+	cmd.AddCommand(tagRemoveCommand())
+
+	return cmd
+}
+
+func tagSetCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:        "set",
+		Short:      "Create or modify revision tags",
+		Example:    "istioctl x tag set prod --revision 1-8-0",
+		SuggestFor: []string{"create"},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 0 {
+				return fmt.Errorf("must provide a tag for modification")
+			}
+			if len(args) > 1 {
+				return fmt.Errorf("can only provide a single tag for creation")
+			}
+
+			client, err := kubeClient(kubeconfig, configContext)
+			if err != nil {
+				return fmt.Errorf("failed to create kubernetes client: %v", err)
+			}
+
+			return setTag(context.Background(), client.Kube(), args[0], revision, cmd.OutOrStdout())
+		},
+	}
+
+	cmd.PersistentFlags().BoolVar(&overwrite, "overwrite", false, "whether to overwrite an existing tag")
+	cmd.PersistentFlags().BoolVarP(&skipConfirmation, "skip-confirmation", "y", false, "whether to skip confirmation for tag deletion")
+	cmd.PersistentFlags().StringVarP(&revision, "revision", "r", "", "revision to point tag to")
+	_ = cmd.MarkPersistentFlagRequired("revision")
+
+	return cmd
+}
+
+func tagListCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "list",
+		Short:   "List existing revision tags and their corresponding revisions",
+		Example: "istioctl x tag list",
+		Aliases: []string{"show"},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) != 0 {
+				return fmt.Errorf("tag list command does not accept arguments")
+			}
+
+			client, err := kubeClient(kubeconfig, configContext)
+			if err != nil {
+				return fmt.Errorf("failed to create kubernetes client: %v", err)
+			}
+			return listTags(context.Background(), client.Kube(), cmd.OutOrStdout())
+		},
+	}
+
+	return cmd
+}
+
+func tagRemoveCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "remove",
+		Short:   "Remove an existing revision tag",
+		Example: "istioctl x tag remove prod",
```

Comment:
Again a longer Example and a Long.  The text should mention that the behavior is undefined if you remove a tag that is currently being used by a namespace istio.io/rev=.